src/backend/scope.rs

```diff
@@ -47,125 +47,147 @@ impl Symbol {
-            list.push(value);
-        } else if let Some(father) = &self.father {
-            father.borrow_mut().symbol_list_append(id, value)?;
-        }
-        Ok(())
+    pub fn symbol_crvalue_list_item(&self, id: usize, index: usize) -> Result<CrValue> {
+        self.get_var(id, |sym| {
+            sym.and_then(Symbol::get_value)
+                .and_then(CrValue::as_list)
+                .map(|list| list[index].clone())
+        })
     }
 
     #[inline]
-    pub fn symbol_list_insert(&mut self, id: &usize, index: usize, value: CrValue) -> Result<()> {
-        if let Some(symbol) = self.symbols.get_mut(id) {
-            let list = symbol.get_value_mut()?.as_list_mut()?;
-            list.insert(index, value);
-        } else if let Some(father) = &self.father {
-            father.borrow_mut().symbol_list_insert(id, index, value)?;
-        }
-        Ok(())
+    pub fn symbol_assign(&mut self, id: usize, value: CrValue) -> Result<()> {
+        self.get_var_mut(id, |sym| sym.and_then(|sym| sym.assign(value)))
     }
 
     #[inline]
-    pub fn symbol_list_modify(&mut self, id: &usize, index: usize, value: CrValue) -> Result<()> {
-        if let Some(symbol) = self.symbols.get_mut(id) {
-            let list = symbol.get_value_mut()?.as_list_mut()?;
-            list[index] = value;
-        } else if let Some(father) = &self.father {
-            father.borrow_mut().symbol_list_modify(id, index, value)?;
-        }
-        Ok(())
+    fn symbol_list_mut(&mut self, id: usize) -> Result<&mut Vec<CrValue>> {
+        self.get_var_mut(id, |sym| {
+            sym.and_then(Symbol::get_value_mut)
+                .and_then(CrValue::as_list_mut)
+        })
     }
 
     #[inline]
-    pub fn symbol_list_remove(&mut self, id: &usize, index: usize) -> Result<CrValue> {
-        if let Some(symbol) = self.symbols.get_mut(id) {
-            let list = symbol.get_value_mut()?.as_list_mut()?;
-            Ok(list.remove(index))
-        } else if let Some(father) = &self.father {
-            Ok(father.borrow_mut().symbol_list_remove(id, index)?)
-        } else {
-            Err(Error::SymbolNotFound)
-        }
+    pub fn symbol_list_append(&mut self, id: usize, value: CrValue) -> Result<()> {
+        self.symbol_list_mut(id).map(|vec| vec.push(value))
+    }
+
+    #[inline]
+    pub fn symbol_list_insert(&mut self, id: usize, index: usize, value: CrValue) -> Result<()> {
+        self.symbol_list_mut(id).map(|vec| vec.insert(index, value))
+    }
+
+    #[inline]
+    pub fn symbol_list_modify(&mut self, id: usize, index: usize, value: CrValue) -> Result<()> {
+        self.symbol_list_mut(id).map(|vec| vec[index] = value)
+    }
+
+    #[inline]
+    pub fn symbol_list_remove(&mut self, id: usize, index: usize) -> Result<CrValue> {
+        self.get_var_mut(id, |sym| {
+            sym.and_then(Symbol::get_value_mut)
+                .and_then(CrValue::as_list_mut)
+                .map(|list| list.remove(index))
+        })
```

Comment:
What was I thinking at the time?
It seems like we should use `symbol_list_mut` here

Reply:
Welcome to make a new pr!